ledger/participant-integration-api/src/main/scala/platform/store/cache/StateCache.scala

```diff
@@ -43,27 +48,45 @@ private[platform] case class StateCache[K, V](cache: Cache[K, V], registerUpdate
     * this method updates the cache only if the to-be-inserted tuple is the most recent
     * (i.e. it has `validAt` highest amongst the competing updates).
     *
+    * NOTE: Calls to `put` evolve the cache index. Callers should ensure strictly increasing validAt evolution.
+    *
     * @param key the key at which to update the cache
     * @param validAt ordering discriminator for pending updates for the same key
     * @param value the value to insert
     */
-  def put(key: K, validAt: Long, value: V): Unit = Timed.value(
-    registerUpdateTimer, {
-      pendingUpdates.synchronized {
-        val competingLatestForKey =
-          pendingUpdates
-            .get(key)
-            .map { pendingUpdate =>
-              val oldLatestValidAt = pendingUpdate.latestValidAt
-              pendingUpdate.latestValidAt = Math.max(validAt, pendingUpdate.latestValidAt)
-              oldLatestValidAt
-            }
-            .getOrElse(Long.MinValue)
+  def put(key: K, validAt: Long, value: V)(implicit loggingContext: LoggingContext): Unit =
+    Timed.value(
+      registerUpdateTimer, {
+        pendingUpdates.synchronized {
+          // Synchronous updates should generally increase the _cacheIndex strictly monotonic.
+          // However, on transient DB errors, the restart of the update flow can lead to
```

Comment:
Good question. My assumption is that once an error is encountered, it is forwarded downstream, cancelling all operators sequentially. And after that it appears on the restart logic, which handles it.

Reply:
> And if it is guaranteed, then I do not understand how can DB errors/resubscription lead to having a _cacheIndex behind resubscription index, since for me it seems they are set in one single synchronous computation.

Here I was not sure and preferred being defensive. I don't know whether the propagated error cancels the operator immediately or allows full execution of it.